catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/authentication/kerberos/HiveBackendProxy.java

```diff
@@ -0,0 +1,137 @@
+public class HiveBackendProxy implements MethodInterceptor {
+
+  private final HiveCatalog target;
+  private final Class<? extends HiveCatalog> targetClazz;
+  private final String kerberosRealm;
+  private final UserGroupInformation proxyUser;
+  private final Options options;
+  private final ClientPool<IMetaStoreClient, TException> newClientPool;
+
+  public HiveBackendProxy(Options options, HiveCatalog target, String kerberosRealm) {
+    this.target = target;
+    targetClazz = target.getClass();
+    this.options = options;
+    this.kerberosRealm = kerberosRealm;
+    try {
+      proxyUser = UserGroupInformation.getCurrentUser();
+
+      // Replace the original client pool with PaimonHiveCachedClientPool. Why do we need to do
+      // this? Because the original client pool in Paimon uses a fixed username to create the
+      // client pool, and it will not work with kerberos authentication. We need to create a new
+      // client pool with the current user. For more, please see CachedClientPool#clientPool and
+      // notice the value of `key`
+      this.newClientPool = resetPaimonHiveCachedClientPool();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to get current user", e);
+    } catch (IllegalAccessException | NoSuchFieldException e) {
+      throw new RuntimeException("Failed to reset PaimonHiveCachedClientPool", e);
+    }
+  }
+
+  @Override
+  public Object intercept(Object o, Method method, Object[] objects, MethodProxy methodProxy)
+      throws Throwable {
+
+    String proxyKerberosPrincipalName = PrincipalUtils.getCurrentPrincipal().getName();
+    if (!proxyKerberosPrincipalName.contains("@")) {
+      proxyKerberosPrincipalName =
+          String.format("%s@%s", proxyKerberosPrincipalName, kerberosRealm);
+    }
+
+    UserGroupInformation realUser =
+        UserGroupInformation.createProxyUser(proxyKerberosPrincipalName, proxyUser);
+
+    String token =
+        newClientPool.run(
+            client ->
+                client.getDelegationToken(
+                    PrincipalUtils.getCurrentPrincipal().getName(), proxyUser.getShortUserName()));
+
+    Token<DelegationTokenIdentifier> delegationToken = new Token<>();
+    delegationToken.decodeFromUrlString(token);
+    realUser.addToken(delegationToken);
+
+    return realUser.doAs(
+        (PrivilegedExceptionAction<Object>)
+            () -> {
+              try {
+                return methodProxy.invoke(target, objects);
+              } catch (Throwable e) {
+                if (RuntimeException.class.isAssignableFrom(e.getClass())) {
+                  throw (RuntimeException) e;
+                }
+                throw new RuntimeException("Failed to invoke method", e);
+              }
+            });
+  }
+
+  private ClientPool<IMetaStoreClient, TException> resetPaimonHiveCachedClientPool()
+      throws IllegalAccessException, NoSuchFieldException {
+    final Field m = HiveCatalog.class.getDeclaredField("clients");
```

Comment:
HiveCatalog‘s clients field may have alive threads. Do you need to close them first? cc @yuqi1129

Reply:
I think so, the same goes for Iceberg Hive backend. The problem is that it's not so easy to close the pool and I have tried to do it in the Iceberg Hive backend but failed.